### PR TITLE
Extend partition metadata (readonly status, statistics)

### PR DESCRIPTION
**Pull request by Jmgr**
This PR adds the readonly status in the partition metadata, as well as various statistics.

### CODE REVIEW

**Comment by Jmgr** [comment]
I'm a bit worried that this might clash with https://github.com/liftbridge-io/liftbridge/issues/222, and whether those metrics should be made accessible though metadata or by some other way. I think I would like to re-create this PR upstream to get some feedback from Tyler first.

**Comment by cruickshankpg** [comment]
> I'm a bit worried that this might clash with liftbridge-io/liftbridge#222

I think these are very different use cases so I'm not too worried. The metrics is for monitoring etc whereas this is very much for a functional purpose.

File: api.proto [hunk 1]
```diff
@@ -214,21 +214,33 @@ message StreamMetadata {
...
-    map<int32, PartitionMetadata> partitions = 4; // Information for the stream partitions
+    string                        name              = 1; // The name of the stream being described
+    string                        subject           = 2; // The stream subject
+    Error                         error             = 3; // Indicates if there was something wrong with the requested stream
+    map<int32, PartitionMetadata> partitions        = 4; // Information for the stream partitions
+    int64                         creationTimestamp = 5 [jstype=JS_STRING]; // When the stream has been created
+}
+
+// PartitionEventMetrics contains metrics related to a stream partition.
+message PartitionEventMetrics {
```
Comment (cruickshankpg) [inline api.proto]: I wouldn't call these metrics, maybe call them `PartitionState` instead. With metric in the name people are at risk of confusing them with the proper metrics work

File: api.proto [hunk 2]
```diff
@@ -214,21 +214,33 @@ message StreamMetadata {
...
+    repeated string       replicas          = 3; // Broker ids of the partition replicas
+    repeated string       isr               = 4; // Broker ids of the in-sync replica set
+    int64                 highWatermark     = 5; // High watermark of the partition
+    int64                 newestOffset      = 6; // Newest offset of the partition
+    bool                  paused            = 7; // Is this partition paused?
+    bool                  readonly          = 8; // Is this partition read-only?
+    int64                 creationTimestamp = 9 [jstype=JS_STRING]; // When the partition has been created
+    PartitionEventMetrics messageMetrics    = 10; // Metrics for received messages
+    PartitionEventMetrics pauseMetrics      = 11; // Metrics for paused status
+    PartitionEventMetrics readonlyMetrics   = 12; // Metrics for readonly status
```
Comment (cruickshankpg) [inline api.proto]: I'm confused what PartitionEventMetrics mean for pause and readonly. Are they when the event was processed?

Comment (Jmgr) [reply api.proto]: Yes, their are only used to store the first and last time the pause and readonly status has been changed. They could also contain the number of times this change has been done, if we believe that could be a useful information.

File: api.proto (hunk 1 above)
Comment (Jmgr) [reply api.proto]: I agree, but I'm not sure about a new name. I feel `State` implies the current state, so `pauseState` or `readonlyState`, but what's stored here is not really the state but rather some historical value about it. `PartitionEventHistory` seems a bit too much. `PartitionEventTimestamps`?

File: api.proto (hunk 2 above)
Comment (cruickshankpg) [reply api.proto]: might be worth a comment giving that context

File: api.proto (hunk 1 above)
Comment (cruickshankpg) [reply api.proto]: yeah `PartitionEventTimestamps` sounds good

Comment (Jmgr) [reply api.proto]: Fixed in 143245f.

File: api.proto (hunk 2 above)
Comment (Jmgr) [reply api.proto]: Fixed in 143245f.

**Comment by Jmgr** [comment]
PR created upstream https://github.com/liftbridge-io/liftbridge-api/pull/41.

**Pull request closed without merging**